Tribler/community/multichain/community.py

```diff
@@ -0,0 +1,427 @@
+        :param message:
+        """
+        block = DatabaseBlock.from_signature_response_message(message)
+        self.logger.info("Persisting sr: %s" % base64.encodestring(block.hash_requester).strip())
+        self.persistence.update_block_with_responder(block)
+
+    def persist_signature_request(self, message):
+        """
+        Persist the signature request message as a block.
+        The block will be updated when a response is received.
+        :param message:
+        """
+        block = DatabaseBlock.from_signature_request_message(message)
+        self.logger.info("Persisting sr: %s" % base64.encodestring(block.hash_requester).strip())
+        self.persistence.add_block(block)
+
+    def send_crawl_request(self, candidate, sequence_number=None):
+        if sequence_number is None:
+            sequence_number = self.persistence.get_latest_sequence_number(candidate.get_member().public_key)
+        self.logger.info("Crawler: Requesting crawl from node %s, from sequence number %d" %
+                         (base64.encodestring(candidate.get_member().mid).strip(), sequence_number))
+        meta = self.get_meta_message(CRAWL_REQUEST)
+        message = meta.impl(authentication=(self.my_member,),
+                            distribution=(self.claim_global_time(),),
+                            destination=(candidate,),
+                            payload=(sequence_number,))
+        self.dispersy.store_update_forward([message], False, False, True)
+
+    def received_crawl_request(self, messages):
+        for message in messages:
+            self.logger.info("Crawler: Received crawl request from node %s, from sequence number %d" %
+                             (base64.encodestring(message.candidate.get_member().mid).strip(),
+                              message.payload.requested_sequence_number))
+            self.crawl_requested(message.candidate, message.payload.requested_sequence_number)
+
+    def crawl_requested(self, candidate, sequence_number):
+        blocks = self.persistence.get_blocks_since(self._public_key, sequence_number)
+        if len(blocks) > 0:
+            self.logger.info("Crawler: Sending %d blocks" % len(blocks))
+            messages = [self.get_meta_message(CRAWL_RESPONSE)
+                        .impl(authentication=(self.my_member,),
+                              distribution=(self.claim_global_time(),),
+                              destination=(candidate,),
+                              payload=block.to_payload()) for block in blocks]
+            self.dispersy.store_update_forward(messages, False, False, True)
+            if len(blocks) > 1:
+                # we sent more than 1 block. Send a resumption token so the other side knows it should continue crawling
+                # last_block = blocks[-1]
+                # resumption_number = last_block.sequence_number_requster if
+                # last_block.mid_requester == self._mid else last_block.sequence_number_responder
```

Comment:
There is some commented code (in more places that this), that allows resumption tokens to indicate where to resume. I'm still debating if this choice should be made by the crawler or the crawlee. For now it is fine to remove this code, I can look it up if i have to.